src/main/kotlin/org/rust/lang/doc/lexer/RustDocHighlightingLexer.flex

```diff
@@ -0,0 +1,190 @@
+package org.rust.lang.doc.lexer;
+
+import com.intellij.lexer.*;
+import com.intellij.psi.tree.IElementType;
+
+import static org.rust.lang.doc.psi.RustDocElementTypes.*;
+import static com.intellij.psi.TokenType.*;
+
+%%
+
+%{
+  private int MAIN_STATE = YYINITIAL;
+
+  public _RustDocHighlightingLexer(boolean isBlock) {
+    this((java.io.Reader)null);
+    MAIN_STATE = isBlock ? IN_BLOCK : IN_EOL;
+  }
+
+  // FIXME(jajakobyly): I think it is possible this will break, though I couldn't make it do so :-)
+  private char CODE_FENCE_DELIM = '\0';
```

Comment:
Ah, there actually can be something funny with `getCurrentPosition` / `restore` pair of methods. Perhaps you should override `getCurrentPosition` to also contain the `CODE_FENCE_DELIM`. 

And perhaps the same should be done for `rust.flex`? Are we [violating the contract](http://www.jetbrains.org/intellij/sdk/docs/reference_guide/custom_language_support/implementing_lexer.html)

>A lexer that can be used incrementally may need to return its state, which means the context corresponding to each position in a file. For example, a Java lexer could have separate states for top level context, comment context and string literal context. An important requirement for a syntax highlighting lexer is that its state must be represented by a single integer number returned from Lexer.getState(). That state will be passed to the Lexer.start() method, along with the start offset of the fragment to process, when lexing is resumed from the middle of a file. Lexers used in other contexts can always return 0 from the getState() method.

? Looks like it, but I have never observed any problems...